### PR TITLE
[main][feat] liquidation reader

### PR DESCRIPTION
**Pull request by kangdaiyu**

### CODE REVIEW

File: src/readers/LiquidationReader.sol [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import { IPerpStorage } from "@hmx/storages/interfaces/IPerpStorage.sol";
+import { ICalculator } from "@hmx/contracts/interfaces/ICalculator.sol";
+
+// libs
+import { SqrtX96Codec } from "@hmx/libraries/SqrtX96Codec.sol";
+import { TickMath } from "@hmx/libraries/TickMath.sol";
+
+contract LiquidationReader {
+  IPerpStorage public perpStorage;
+  ICalculator public calculator;
```
Comment (john-xina-p88) [inline src/readers/LiquidationReader.sol:17]:
```suggestion
  IPerpStorage immutable public perpStorage;
  ICalculator immutable public calculator;
```
gas saving

File: src/readers/LiquidationReader.sol [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    bytes32[] memory _assetIds,
+    uint64[] memory _prices,
+    bool[] memory _shouldInverts
+  ) external view returns (address[] memory) {
+    address[] memory subAccounts = perpStorage.getActiveSubAccounts(_limit, _offset);
+
+    uint256[] memory prices;
+    uint256 len = _prices.length;
+    prices = new uint256[](len);
+    for (uint256 i = 0; i < len; i++) {
```
Comment (john-xina-p88) [inline src/readers/LiquidationReader.sol:36]:
```suggestion
    for (uint256 i; i < len; ) {
    
    ...
    unchecked {
      ++i;
    }
```
gas saving

File: src/readers/LiquidationReader.sol [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      if (_assetIds[i] == bytes32(abi.encodePacked("GLP"))) {
+        prices[i] = uint256(_prices[i]) * 1e22;
+        continue;
+      }
+      prices[i] = _convertPrice(_prices[i], _shouldInverts[i]);
+    }
+
+    len = subAccounts.length;
+    address[] memory liquidatableSubAccounts = new address[](len);
+    for (uint256 i = 0; i < len; i++) {
```
Comment (john-xina-p88) [inline src/readers/LiquidationReader.sol:46]:
```suggestion
    for (uint256 i; i < len; ) {
    
    ...
    unchecked {
      ++i;
    }
```
gas saving

File: src/readers/LiquidationReader.sol [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  constructor(address _perpStorage, address _calculator) {
+    perpStorage = IPerpStorage(_perpStorage);
+    calculator = ICalculator(_calculator);
+  }
+
+  function getLiquidatableSubAccount(
+    uint64 _limit,
+    uint64 _offset,
+    bytes32[] memory _assetIds,
+    uint64[] memory _prices,
```
Comment (kim-hera) [inline src/readers/LiquidationReader.sol:28]: Please add the spec of this price ie, e8? flip? inverted? etc

File: src/readers/LiquidationReader.sol [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    bool[] memory _shouldInverts
+  ) external view returns (address[] memory) {
+    address[] memory subAccounts = perpStorage.getActiveSubAccounts(_limit, _offset);
+
```
Comment (kim-hera) [inline src/readers/LiquidationReader.sol:32]: Add the validation of _assetIds, _prices, _shouldInverts length.

**Review by john-xina-p88 — APPROVED** [review]

File: src/readers/LiquidationReader.sol [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  /// @notice Get the liquidatable sub-accounts.
+  /// @param _limit The maximum number of sub-accounts to retrieve.
+  /// @param _offset The offset for fetching sub-accounts.
+  /// @param _assetIds An array of asset IDs.
+  /// @param _pricesE8 An array of prices in E8 format corresponding to the asset IDs.
+  /// @param _shouldInverts An array of boolean values indicating whether to invert prices for the asset IDs.
+  /// @return liquidatableSubAccounts An array of sub-account addresses that meet the liquidation criteria.
+  function getLiquidatableSubAccount(
+    uint64 _limit,
+    uint64 _offset,
```
Comment (kim-hera) [inline src/readers/LiquidationReader.sol:35]: Limit, offset is not used as pagination of "liquidatable subaccount", but "active sub account" instead. Not sure if this is an intention.

**Review by kim-hera — APPROVED** [review]

**Pull request merged**